delegates/src/main/java/com/revolut/recyclerkit/delegates/DiffAdapter.kt

```diff
@@ -25,22 +30,144 @@ import androidx.recyclerview.widget.DiffUtil
+    async: Boolean = true,
+    autoScrollToTop: Boolean = false
 ) : AbsRecyclerDelegatesAdapter(delegatesManager) {
 
-    private val differ: AsyncListDiffer<ListItem> = AsyncListDiffer(
-        AdapterListUpdateCallback(this),
-        AsyncDifferConfig.Builder(ListDiffCallback<ListItem>()).build()
-    )
+    protected interface DifferDelegate {
+        val items: List<ListItem>
+        fun attachRecyclerView(recyclerView: RecyclerView)
+        fun detachRecyclerView(recyclerView: RecyclerView)
+        fun setItems(items: List<ListItem>)
+    }
+
+    private val differDelegate = if (async) {
+        AsyncDifferStrategy(adapter = this, autoScrollToTop = autoScrollToTop)
+    } else {
+        SyncDifferStrategy(adapter = this, autoScrollToTop = autoScrollToTop, detectMoves = true)
+    }
+
+    open val items: List<ListItem>
+        get() = differDelegate.items
+
+    override fun getItem(position: Int): ListItem = items[position]
+
+    @UiThread
+    open fun setItems(items: List<ListItem>) {
+        check(Looper.myLooper() == Looper.getMainLooper()) { "DiffAdapter.setItems() was called from worker thread" }
+        differDelegate.setItems(items)
+    }
+
+    override fun getItemCount() = items.size
+
+    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
+        differDelegate.attachRecyclerView(recyclerView)
+    }
 
-    override fun getItem(position: Int): ListItem = differ.currentList[position]
+    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
+        differDelegate.detachRecyclerView(recyclerView)
+    }
 
-    fun setItems(items: List<ListItem>) = differ.submitList(items)
+    private class AsyncDifferStrategy(
+        adapter: RecyclerView.Adapter<*>,
+        private val autoScrollToTop: Boolean
+    ) : DifferDelegate {
+        private val differ: AsyncListDiffer<ListItem> = AsyncListDiffer(
+            AdapterListUpdateCallback(adapter),
+            AsyncDifferConfig.Builder(ListDiffItemCallback<ListItem>()).build()
+        )
+        private var recyclerView = WeakReference<RecyclerView>(null)
+        override val items: List<ListItem>
+            get() = differ.currentList
 
-    override fun getItemCount() = differ.currentList.size
+        override fun attachRecyclerView(recyclerView: RecyclerView) {
+            this.recyclerView = WeakReference(recyclerView)
+        }
 
-    private class ListDiffCallback<T> : DiffUtil.ItemCallback<ListItem>() {
+        override fun detachRecyclerView(recyclerView: RecyclerView) {
+            this.recyclerView = WeakReference(null)
+        }
+
+        override fun setItems(items: List<ListItem>) {
+            val recyclerViewRef = recyclerView
+            val rv = recyclerViewRef.get() ?: error("Recycler View not attached")
+            val layoutManager = rv.layoutManager
+
+            val firstVisiblePosition = if (autoScrollToTop && layoutManager is LinearLayoutManager) {
```

Comment:
There is code duplication for autoScroll logic in SyncDifferStrategy and AsyncSyncDifferStrategy.
- Let's put code into lamda and use it in both cases.
- Let's invoke it always after submitting list to recycler view.  Now it happens in different moments